src/pkg/caendr/caendr/models/job_pipeline/heritability_pipeline.py

```diff
@@ -80,31 +93,107 @@ def parse(cls, data, valid_file_extensions=None):
       raise DataFormatError("The heritability data contain fewer than five unique strains. Please be sure to measure trait values for at least five wild strains in at least three independent assays.")
 
     # Compute hash from file
-    data_hash = get_file_hash(local_path, length=32)
+    data_hash = get_file_hash(local_file, length=32)
 
     # Extract trait from table data
     data['trait'] = csv_row[2]
 
     return {
       'props': data,
       'hash':  data_hash,
-      'files': [local_path],
+      'files': [local_file],
     }
 
 
+
+  #
+  # Fetching & Parsing Input
+  #
+
+  def _parse_input(self, blob):
+
+    # Download data
+    data = download_blob_as_dataframe(blob)
+
+    # If data file is empty, raise an error
+    if data is None or data.empty:
+      raise EmptyReportDataError(self.report.id)
+
+    # Parse data file
+    data['AssayNumber'] = data['AssayNumber'].astype(int)
+    data['label'] = data.apply(lambda x: f"{x['AssayNumber']}: {x['Value']}", 1)
+
+    # Return parsed data
+    return data.to_dict('records')
+
+
+
   #
-  # File Storage
+  # Fetching & Parsing Output
   #
 
-  def upload(self, data_files):
 
-    # Heritability only takes one upload file
-    if len(data_files) > 1:
-      raise ValueError('Only one data file should be uploaded.')
+  def fetch_output(self, raw: bool = False):
+    result = super().fetch_output(raw=raw)
+
+    # If result exists and report status hasn't been marked as finished yet, update it
+    if result is not None and not self.is_finished():
```

Comment:
Moved out of the `view` function, so that from the caller's perspective this is handled implicitly